### ISSUE TITLE
Overlapping computations / prefetching like Tensorflow dataset pipeline?

### ISSUE THREAD

**Issue by Hoeze** [dask/dask#5602]
**TL;DR:** Is there a way to speed up fetching sub-batches of the same chunk?

Hi, I'm trying to feed a stacked array of size (134,033,856 x 4) with a chunk size of (315,230 x 1) into Keras.
Therefore I wrote the following generator:
```python
def dask_batch_iter(data, drop_last=False, batch_size=None):
    if batch_size == None:
        batch_size = data.chunksize[0]
    num_samples = np.shape(data)[0]
    num_batches = num_samples // batch_size

    for i in range(num_batches):
        batch_start = i * batch_size
        yield dask.compute(data[batch_start: batch_start + batch_size])[0]
    
    if not drop_last and num_samples % batch_size != 0:
        yield dask.compute(data[num_batches * batch_size:])[0]
```
I noted that fetching batches of size 1,000 takes about 100x the training time than fetching batches of size 315,230. This suggests that Dask re-computes the whole chunk each time I fetch a 1,000 samples batch from it.

As a workaround, I tried dask.cache:
```python
from dask.cache import Cache
cache = Cache(8e9)  # Leverage eight gigabytes of memory
cache.register()
```
This does now cache all the Zarr chunks, but not the four resulting (315,230 x 1) chunks. Its speedup is therefore very low.

What I'd need would be a way to tell Dask that I will fetch additional batches from the same chunk and therefore it should always keep the last (315,230 x 4) block. E.g. this could be solved with some job queue similar to Tensorflows Dataset API.
I manually wrote some code which fetches a chunk and returns parts of it as batches, but my use case is a quite common pattern which should be somehow supported by dask IMHO.

**Comment by mrocklin** [comment]
If you want to use dask array the you are probably looking for the persist
method.

If you want to be higher performance and do overlapping then you might want
to look at the concurrent.futures API.
https://docs.dask.org/en/latest/futures.html

**Comment by Hoeze** [comment]
Thanks for your answer @mrocklin.
The `persist` method does not really look like what I want, because it loads the whole dataset into memory. Instead, I only want to load the last chunk into memory.

Futures do look more like what I'd like, but does this concept work in a single process?
Otherwise Dask would do a lot of memcopy, wouldn't it?

**Comment by TomAugspurger** [comment]
@Hoeze you can start a distributed LocalCluster using just threads with `processes=False` https://distributed.dask.org/en/latest/local-cluster.html#distributed.deploy.local.LocalCluster.

If you have multiple machines / processes, yes: chunks of the data will need to be moved around. Dask tries to minimize that.

**Comment by Hoeze** [comment]
Wow, thanks @TomAugspurger. I did not know about this one.
